emperor/support_files/js/color-view-controller.js

```diff
@@ -193,71 +204,154 @@ define([
       return [colors, gradientSVG];
     }
 
-    if (discrete === false) {
+    //Call helper function to create the required colormap type
+    if (discrete) {
+      colors = ColorViewController.getDiscreteColors(values, map);
+    } else if (scaled) {
+      var info = ColorViewController.getScaledColors(values, map);
+      colors = info[0];
+      gradientSVG = info[1];
+    } else {
+      colors = ColorViewController.getInterpolatedColors(values, map)
+    }
+    return [colors, gradientSVG];
+  };
+
+  /**
+   *
+   * Retrieve a discrete color.
+   *
+   * @param {values} list of objects to generate a color for, usually a
+   * category in a given metadata column.
+   * @param {map} string, name of the discrete color map to use.
+   *
+   * This function will generate a list of coloring values depending on the
+   * coloring scheme that the system is currently using (discrete or
+   * continuous).
+   *
+   * Defaults to use qiime discrete colors if there's no map passed in.
+   *
+   */
+  ColorViewController.getDiscreteColors = function(values, map) {
+    var map = map || 'discrete-coloring-qiime';
+
+    if (map == 'discrete-coloring-qiime') {
+      map = ColorViewController._qiimeDiscrete;
+    } else {
       map = chroma.brewer[map];
-      interpolator = chroma.bezier([map[0], map[3], map[4], map[5], map[8]]);
-      if (scaled === true) {
-        // Assuming we have numeric since scaled called
-        var numericValues = _.map(values, Number);
-        min = _.min(numericValues);
-        max = _.max(numericValues);
-        interpolator = chroma.scale(map).domain([min, max]);
-      }
     }
+    var size = map.length;
+    var colors = {};
+    for (var i = 0; i < values.length; i++) {
+        mapIndex = i - (Math.floor(i / size) * size);
+        colors[values[i]] = map[mapIndex];
+    }
+    return colors;
+  };
 
-    for (var index in values) {
-      if (discrete) {
-        // get the next available color
-        colors[values[index]] = ColorViewController.getDiscreteColor(index, map);
-      }
-      else if (scaled === true) {
-        colors[values[index]] = interpolator(Number(values[index])).hex();
+  /**
+   *
+   * Retrieve a scaled color set with scale bar.
+   *
+   * @param {values} list of objects to generate a color for, usually a
+   * category in a given metadata column.
+   * @param {map} string, name of the discrete color map to use.
+   *
+   * This function will generate a list of coloring values depending on the
+   * coloring scheme that the system is currently using (discrete or
+   * continuous).
+   *
+   * Defaults to use viridis colormap if there's no map passed in.
+   *
+   */
+  ColorViewController.getScaledColors = function(values, map, nanColor) {
+    var map = map || 'viridis';
+    var nanColor = nanColor || '#64655d';
+    map = chroma.brewer[map];
+
+    // Get list of only numeric values, error if none
+    var numericValues = values.filter(function(v) { return !isNaN(v)});
```

Comment:
One way to do it would be to do a single scan where we search and count
things, if we could also sort as we are doing this, that would be
exactly what we want. Maybe we can sketch something out.

On (May-09-16| 8:46), Joshua Shorenstein wrote:
> > +   * @param {map} string, name of the discrete color map to use.
> > +   *
> > +   * This function will generate a list of coloring values depending on the
> > +   * coloring scheme that the system is currently using (discrete or
> > +   * continuous).
> > +   *
> > +   * Defaults to use viridis colormap if there's no map passed in.
> > +   *
> > +   */
> > +  ColorViewController.getScaledColors = function(values, map, nanColor) {
> > +    var map = map || 'viridis';
> > +    var nanColor = nanColor || '#64655d';
> > +    map = chroma.brewer[map];
> > +
> > +    // Get list of only numeric values, error if none
> > +    var numericValues = values.filter(function(v) { return !isNaN(v)});
> 
> The two passes is to do the validation that we do indeed have enough values to do the scaling. Not sure it can be done without two passes.
> 
> ---
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/emperor/pull/468/files/c5c2d10589eb2a5c7dd4d326dfa32745e0ee82cc#r62522352